packages/cloud_firestore/lib/src/firestore.dart

```diff
@@ -179,9 +181,11 @@ class FirestoreMessageCodec extends StandardMessageCodec {
       case _kDateTime:
         return new DateTime.fromMillisecondsSinceEpoch(buffer.getInt64());
       case _kGeoPoint:
-        return new GeoPoint(readValue(buffer), readValue(buffer));
+        return new GeoPoint(buffer.getFloat64(), buffer.getFloat64());
       case _kDocumentReference:
-        return Firestore.instance.document(readValue(buffer));
+        final int length = readSize(buffer);
+        final String path = utf8.decoder.convert(buffer.getUint8List(length));
+        return Firestore.instance.document(path);
```

Comment:
Looking at the changes you had to do to replace `read/writeValue` with directly typed alternatives, it becomes painfully clear to me that the standard codec APIs could be much more consistent across Dart, Java, and ObjC. https://github.com/flutter/flutter/issues/15779

I don't think that should hold this PR back.

Reply:
I already had the code that did not use `read/writeValue` from the original version of PR. I had switched to improve readability, but I hadn't considered the excess step of processing the type and the extra bytes in the buffer. This style with known types will be marginally faster.